File: week8&9/Ariling/boj-11265-끝나지 않는 파티/main.py
import sys
input = sys.stdin.readline
import heapq

N, M = map(int, input().split())
dist_info = [list(map(int, input().split())) for _ in range(N)]
user_info = [list(map(int, input().split())) for _ in range(M)]
inf = int(1e9)
graph = [[] for _ in range(N+1)]
answer_info = [[]]

for i in range(N):
    for j in range(N):
        if i == j:
            continue
        else:
            graph[i+1].append((j+1, dist_info[i][j]))
def dikstra(i):
    distances = [inf] * (N+1)
    q = []
    distances[i] = 0
    # 거리를 먼저 넣어야 하는구나..!
    heapq.heappush(q, (0, i))
    while q:
        dis, now = heapq.heappop(q)
        if distances[now] < dis:
            continue
        for to, cost in graph[now]:
            dist = dis + cost
            if dist < distances[to]:
                distances[to] = dist
                heapq.heappush(q, (dist, to))
    return distances
for i in range(1, N+1):
    answer_arr = dikstra(i)
    answer_info.append(answer_arr)
for i in range(M):
    [st, ed, time] = user_info[i]
    if answer_info[st][ed] <= time:
        print("Enjoy other party")
    else:
        print("Stay here")


# for k in range(1, N+1):
#     for i in range(1, N+1):
#         for j in range(1, N+1):
#             graph[i][j] = min(graph[i][j], graph[i][k] + graph[k][j])
# for i in range(M):
#     [st, ed, time] = user_info[i]
#     if graph[st][ed] <= time:
#         print("Enjoy other party")
#     else:
#         print("Stay here")

File: week8&9/Ariling/boj-11265-끝나지 않는 파티/main.js
const fs = require("fs");
const [firstLine, ...input] = fs
  .readFileSync(process.platform === "linux" ? "/dev/stdin" : "./input.txt")
  .toString()
  .trim()
  .split("\n");

const [node, guest] = firstLine.split(" ").map(Number);
const graph = Array(node + 1)
  .fill()
  .map(() => Array(node + 1).fill(Infinity));

for (let i = 0; i < node; i++) {
  const info = input[i].split(" ").map(Number);
  for (let j = 0; j < node; j++) {
    graph[i + 1][j + 1] = info[j];
  }
}

for (let k = 1; k <= node; k++) {
  for (let i = 1; i <= node; i++) {
    for (let j = 1; j <= node; j++) {
      if (graph[i][j] > graph[k][j] + graph[i][k]) {
        graph[i][j] = graph[k][j] + graph[i][k];
      }
    }
  }
}

for (let i = 0; i < guest; i++) {
  // 출발, 도착, 시간순
  const [st, ed, time] = input[i + node].split(" ").map(Number);
  if (graph[st][ed] <= time) console.log("Enjoy other party");
  else console.log("Stay here");
}
